src/adapter/src/client.rs

```diff
@@ -385,157 +416,251 @@ impl SessionClient {
+                future: rows,
+                span: _,
+            } => {
+                let rows = match rows.await {
+                    PeekResponseUnary::Rows(rows) => rows,
+                    PeekResponseUnary::Error(e) => {
+                        return Ok(SimpleResult::err(e));
+                    }
+                    PeekResponseUnary::Canceled => {
+                        return Ok(SimpleResult::err("statement canceled due to user request"));
+                    }
+                };
+                let mut sql_rows: Vec<Vec<serde_json::Value>> = vec![];
+                let col_names = match desc.relation_desc {
+                    Some(desc) => desc.iter_names().map(|name| name.to_string()).collect(),
+                    None => vec![],
+                };
+                let mut datum_vec = mz_repr::DatumVec::new();
+                for row in rows {
+                    let datums = datum_vec.borrow_with(&row);
+                    sql_rows.push(datums.iter().map(From::from).collect());
                 }
-                | ExecuteResponse::Updated(_)
-                | ExecuteResponse::AlteredObject(_)
-                | ExecuteResponse::AlteredIndexLogicalCompaction
-                | ExecuteResponse::AlteredSystemConfiguraion
-                | ExecuteResponse::Deallocate { all: _ }
-                | ExecuteResponse::Prepare) => {
-                    results.push(SimpleResult::ok(res));
+                Ok(SimpleResult::Rows {
+                    rows: sql_rows,
+                    col_names,
+                })
+            }
+            ExecuteResponse::Fetch { .. }
+            | ExecuteResponse::SetVariable { .. }
+            | ExecuteResponse::Tailing { .. }
+            | ExecuteResponse::CopyTo { .. }
+            | ExecuteResponse::CopyFrom { .. }
+            | ExecuteResponse::Raise { .. }
+            | ExecuteResponse::DeclaredCursor
+            | ExecuteResponse::ClosedCursor => {
+                // NOTE(benesch): it is a bit scary to ignore the response
+                // to these types of planning *after* they have been
+                // planned, as they may have mutated state. On a quick
+                // glance, though, ignoring the execute response in all of
+                // the above situations seems safe enough, and it's a more
+                // target allowlist than the code that was here before.
+                Ok(SimpleResult::err(
+                    "executing statements of this type is unsupported via this API",
+                ))
+            }
+        }
+    }
+
+    /// Executes an [`HttpSqlRequest`].
+    ///
+    /// The provided `stmts` are executed directly, and their results
+    /// are returned as a vector of rows, where each row is a vector of JSON
+    /// objects.
+    pub async fn execute_sql_http_request(
+        &mut self,
+        request: HttpSqlRequest,
+    ) -> Result<SqlHttpExecuteResponse, AdapterError> {
```

Comment:
Can this leave as a free function in the `environmentd::http` module? This method looks spiritually equivalent to the pgwire server—i.e., a layer on _top_ of the "adapter client" that adapter doesn't actually need to know about. Put another way: I'd love for the `adapter` crate not to have any types/methods with "HTTP" in their name, since it (roughly) manages to be agnostic to any types/methods with "pgwire" in their name.